java-spiffe-core/grpc-netty-macos-aarch64/src/main/java/io/spiffe/workloadapi/internal/GrpcManagedChannelFactory.java

```diff
@@ -0,0 +1,81 @@
+package io.spiffe.workloadapi.internal;
+
+import io.grpc.ManagedChannel;
+import io.grpc.netty.NegotiationType;
+import io.grpc.netty.NettyChannelBuilder;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.kqueue.KQueueDomainSocketChannel;
+import io.netty.channel.kqueue.KQueueEventLoopGroup;
+import io.netty.channel.unix.DomainSocketAddress;
+import lombok.NonNull;
+import lombok.val;
+import org.apache.commons.lang3.SystemUtils;
+
+import java.net.URI;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Factory for creating ManagedChannel instances for Mac OS.
+ */
+public final class GrpcManagedChannelFactory {
+
+    private static final String UNIX_SCHEME = "unix";
+    private static final String TCP_SCHEME = "tcp";
+
+    private GrpcManagedChannelFactory() {
+    }
+
+    /**
+     * Returns a ManagedChannelWrapper that contains a {@link ManagedChannel} to the SPIFFE Socket Endpoint provided.
+     *
+     * @param address         URI representing the Workload API endpoint.
+     * @param executorService the executor to configure the event loop group
+     * @return a instance of a {@link ManagedChannelWrapper}
+     */
+    public static ManagedChannelWrapper newChannel(@NonNull URI address, ExecutorService executorService) {
+        val scheme = address.getScheme();
+        ManagedChannelWrapper result;
+        switch (scheme) {
+            case UNIX_SCHEME:
+                result = createNativeSocketChannel(address, executorService);
+                break;
+            case TCP_SCHEME:
+                result = createTcpChannel(address);
+                break;
```

Comment:
AFAIK the spire-agent does not provide any means to open a tcp socket. What is this for?

Reply:
You are right, spire-agent doesn't, but the library was implemented following the SPIFFE spec that says that the Workload Endpoint may be exposed as either a Unix Domain Socket (UDS) or a TCP socket.